html/semantics/forms/constraints/support/validator.js

```diff
@@ -389,39 +389,54 @@ var validator = {
   },
 
   iterate_over: function(ctl, data) {
-    // Iterate over normal, disabled, readonly, and both.
+    // Iterate over normal, disabled, readonly, and both (if applicable).
+    var ctlNormal = ctl.cloneNode(true);
+    this.set_conditions(ctlNormal, data.conditions);
+    if (data.dirty)
+      this.set_dirty(ctlNormal);
+
     var ctlDisabled = ctl.cloneNode(true);
     this.set_conditions(ctlDisabled, data.conditions);
     if (data.dirty)
       this.set_dirty(ctlDisabled);
     ctlDisabled.disabled = true;
 
-    var ctlReadonly = ctl.cloneNode(true);
-    this.set_conditions(ctlReadonly, data.conditions);
-    if (data.dirty)
-      this.set_dirty(ctlReadonly);
-    ctlReadonly.readonly = true;
+    var expectedImmutable =
+      data.expectedImmutable !== undefined ? data.expectedImmutable : data.expected;
 
-    var ctlBoth = ctl.cloneNode(true);
-    this.set_conditions(ctlBoth, data.conditions);
-    if (data.dirty)
-      this.set_dirty(ctlBoth);
-    ctlBoth.disabled = true;
-    ctlBoth.readonly = true;
-
-    ctl = ctl.cloneNode(true);
-    this.set_conditions(ctl, data.conditions);
-
-    return [
-      {ctl: ctl, data: data, condStr: '.'},
-      {ctl: ctlDisabled, data: data, condStr: ', when control is disabled.'},
-      {ctl: ctlReadonly, data: data, condStr: ', when control is readonly.'},
-      {
+    var variants = [
+      {ctl: ctlNormal, expected: data.expected, condStr: '.'},
+      {ctl: ctlDisabled, expected: expectedImmutable, condStr: ', when control is disabled.'},
+    ];
+
+    if ('readOnly' in ctl) {
```

Comment:
`<input type=checkbox readonly>` supports `readOnly` IDL attribute, but it doesn't make the element immutable.

Reply:
Only textarea and inputs in textual mode have different test expectations for `valueMissing` based on mutability - if element is immutable `valueMissing` should be always false. I introduced `expectedImmutable` to make it possible to specify it in tests.

Maybe instead of introducing `expectedImmutable` it would be better to do this check in `validator.js` in `test_valueMissing` like it's done for `validationMessage`? Then there will be no need to modify test files, only `validator.js`.

Reply:
> Only textarea and inputs in textual mode have different test expectations for `valueMissing` based on mutability 

Ah, you're right.  I thought `<input type=checkbox disabled required>` was NOT suffering from being missing. But it **is** according to https://html.spec.whatwg.org/C/#checkbox-state-(type=checkbox)

> it would be better to do this check in validator.js

I think the code would be complicated because we need to check input types and test_isValid() is also affected.  The current PR looks reasonable as is.